Add DoughnutV0

holygits opened the pull request:
- Add doughnut v0
- Allow runtime to define doughnut verification through DoughnutVerifier trait

Tests compile but there are a few failing in node/executor

xlc commented on core/sr-primitives/Cargo.toml:
```diff
@@ -13,6 +13,7 @@ substrate-primitives = { path = "../primitives", default-features = false }
 rstd = { package = "sr-std", path = "../sr-std", default-features = false }
 runtime_io = { package = "sr-io", path = "../sr-io", default-features = false }
 log = { version = "0.4", optional = true }
+doughnut_rs = { git = "https://github.com/cennznet/doughnut-rs", tag = "v0.1.0", default-features = false }
```
import it as `doughnut`

wilfred-centrality commented on srml/support/src/dispatch.rs:
```diff
@@ -922,16 +924,16 @@ macro_rules! decl_module {
+						use $crate::dispatch::DoughnutVerifier;
+						// Check if a doughnut exists in this execution context and whether it grants permission to
+						// dispatch the call.
+						if let Some(doughnut) = $crate::storage::unhashed::get(b"doughnut") {
```
Let's move the doughnut existential condition and surfacing to a proxy class.
Reasoning: These steps are needed prior to `verify_doughnut` calls, so let's save ourselves some work and write it once.

holygits replied:
After some thinking I don't want to tackle the proxy class until we start with contract integration otherwise it may be throwaway work

xlc commented on core/sr-primitives/src/lib.rs:
```diff
@@ -40,6 +40,10 @@ pub mod testing;
 pub mod transaction_validity;
+pub mod doughnut {
+	//! Runtime doughnut types
+	pub use doughnut_rs::v0::parity::DoughnutV0;
```
I need to check doughnut_rs code, but why there is module called `parity`?

holygits replied:
Parity codec

wilfred-centrality commented on srml/support/src/additional_traits.rs:
```diff
@@ -46,3 +46,23 @@ impl<T, U> ChargeFee<T> for DummyChargeFee<T, U> {
 }
+
+/// A type which can verify doughnuts for an inherent permission domain
+pub trait DoughnutVerifier<Doughnut> {
```
Let's name this a `DoughnutDomainVerifier`.
Reasoning: `DoughnutVerifier` has another meaning within this context.

holygits replied:
I renamed this to `DispatchVerifier`

wilfred-centrality commented on srml/executive/src/lib.rs:
```diff
@@ -265,10 +262,13 @@ impl<
 		// Decode parameters and dispatch
 		let (f, s) = xt.deconstruct();
 		let r = f.dispatch(s.into());
-		storage::unhashed::kill(well_known_keys::DOUGHNUT_KEY);
```
Let's remove the doughnut from storage post-storage.
Reasoning: Without removing it, and if the next extrinsic to execute does not have a doughnut, it will use the previous extrinsic's doughnut.

wilfred-centrality replied:
More robust: add the removal before the `put` logic above (line 264.
Reasoning: This handles all edge-cases besides storage failure.

xlc commented on node/runtime/src/doughnut.rs:
```diff
@@ -0,0 +1,21 @@
+//!
```
add license header

xlc commented on node/runtime/src/doughnut.rs:
```diff
@@ -0,0 +1,21 @@
+		_module: &str,
+		_method: &str,
+	) -> Result<(), &'static str> {
+        // Doughnut verification should always fail in this runtime
```
indent

xlc commented on srml/support/src/additional_traits.rs:
```diff
@@ -46,3 +46,23 @@ impl<T, U> ChargeFee<T> for DummyChargeFee<T, U> {
+		doughnut: &Doughnut,
+		module: &str,
+		method: &str,
+	) -> Result<(), &'static str>;
```
I really don't like all the `&'static str` error string in Substrate.
Can we do something like 
```
enum DoughnutError {
  CustomError(&'static str),
  NotSupported,
  SomeOtherCommonError,
}
```

holygits replied:
Good idea, I might make error an associated type. Since we want runtimes to implement this bit ad-hoc

holygits replied:
I won't do this in this PR because it's going to be more complex than I thought since module calls still use `Result<(), &'static str>`

xlc commented on srml/support/src/dispatch.rs:
```diff
@@ -922,16 +924,16 @@ macro_rules! decl_module {
+							// TO-DOUGH: Eventually we'll want to pass `$param_name` and `$param`
+							let _ = <T as system::Trait>::DoughnutVerifier::verify_doughnut(
+								&doughnut,
+								env!("CARGO_PKG_NAME"), // module
```
in case you don't know, there is `module_path` macro. but I am not sure if it is better than `env!("CARGO_PKG_NAME")` for this case

wilfred-centrality replied:
@xlc It seems in our case `env!` is the more direct solution

holygits commented:
Will leave the remaining pieces of work post-demo

xlc commented on srml/executive/src/lib.rs:
```diff
@@ -265,10 +262,18 @@ impl<
 		}
 
+		// Ensure doughnut storage is clean/reset
+		storage::unhashed::kill(well_known_keys::DOUGHNUT_KEY);
```
kill then put seems redundant
just
```
if ... {
  put(...)
} else {
  kill()
}
```

xlc commented:
This should fixes the test build issues right? In that case I won't spend more time add `Signature` to all the tests

holygits commented:
Yes we removed Signature from srml-system as it wasn't need anymore but added a DoughnutVerifier type to srml-system. Updated tests so that they compile

holygits commented:
Ready for another review

xlc commented on node/primitives/src/plug_extrinsic.rs:
```diff
@@ -41,35 +41,51 @@ fn encode_with_vec_prefix<T: Encode, F: Fn(&mut Vec<u8>)>(encoder: F) -> Vec<u8>
+	/// Doughnut attached, if any
+	pub doughnut: Option<Doughnut>,
+	/// phantom to pipe through AccountId
+	pub phantom: rstd::marker::PhantomData<AccountId>,
```
```suggestion
	pub _phantom: rstd::marker::PhantomData<AccountId>,
```

xlc commented on node/primitives/src/plug_extrinsic.rs:
```diff
@@ -188,45 +201,44 @@ where
-				return Err("invalid doughnut");
+		let verified = if let Some(ref doughnut) = self.doughnut {
+			// TODO: This may need a shim trait e.g. AsRef<bytes>
+			let doughnut_signature = Signature::decode(&mut &doughnut.signature().borrow()[..]).expect("doughnut has a valid signature");
```
`expect` will panic when failed, I think you want `ok_or`

holygits commented:
@cam-cen tests failing because of build script I think

xlc commented:
Caused by this https://github.com/cennznet/plug-blockchain/blob/master/scripts/common.sh#L9

holygits commented:
@xlc can we merge?
